aip/0147.md

```diff
@@ -0,0 +1,81 @@
+---
+aip:
+  id: 147
+  state: draft
+  created: 2020-07-24
+permalink: /147
+redirect_from:
+  - /0147
+---
+
+# Sensitive fields
+
+Sometimes APIs need to collect sensitive information such as private encryption
+keys meant to be _stored_ by the underlying service but not intended to be
+_read_ after writing due to the sensitive nature of the data. For this type of
+data, extra consideration is required for the representation of the sensitive
+data in API requests and responses.
+
+## Guidance
+
+If the sensitive information is _required_ for the resource as a whole to
+exist, the data **should** be accepted as an [input-only field][input-only]
+with no corresponding output field. Because the sensitive data must be present
+for the resource to exist, users of the API may assume that existence of the
+resource implies storage of the sensitive data. For example:
+
+```proto
+message SelfManagedKeypair {
+  string name = 1;
+  // The public key data in PEM-encoded form.
+  bytes public_key = 2;
+  // The private key data in PEM-encoded form.
+  bytes private_key = 3 [
+    (google.api.field_behavior) = INPUT_ONLY,
+  ];
+}
+```
+
+If the sensitive information is _optional_ within the containing resource, an
+[output-only][] boolean field with a postfix of `_set` **should** be used to
+indicate whether or not the sensitive information is present. For example:
+
+```proto
+message Integration {
+  string name = 1;
+  string url = 2;
+  // A secret to be passed in the `Authorization` header of the webhook.
+  string shared_secret = 3 [
+    (google.api.field_behavior) = INPUT_ONLY,
+  ];
+  // True if a `shared_secret` has been set for this Integration.
+  bool shared_secret_set = 4 [
+    (google.api.field_behavior) = OUTPUT_ONLY,
+  ];
```

Comment:
Should these be in a `oneof`? (Honestly unsure.)

Reply:
That's interesting...even though I think you could make the case that they technically comprise a `oneof`, I don't think I'd prefer it.

The semantics of `oneof` are not super-pleasant (at least in Go where I'm most familiar) and the `INPUT_ONLY` and `OUTPUT_ONLY` provide clear enough behavioral direction for servers -- you should always discard/ignore `OUTPUT_ONLY` when reading requests and never populate `INPUT_ONLY` when creating a response.

That said if you think it ought to be that's reasonable to me.

Question (because I don't know): do field behaviors alter behavior of client libraries? Or are they purely informational?

Reply:
> Question (because I don't know): do field behaviors alter behavior of client libraries? Or are they purely informational?

`REQUIRED` does, but the others are informational only at this time. They also alter documentation slightly.

Reply:
I agree that oneofs are pretty horrible in Go, but it still seems semantically correct to me. :-/

Reply:
That said, this is fine to get to reviewing. We can discuss it at approval time.